design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,477 @@
+this proposal is going to make it into implementation.
+
+Since component is becoming mutable, a very common question is how will the component versioning mechanism work.
+
+One of the requirement is the versioning mechanism should allow users to specify a fixed version of component
+in ApplicationConfiguration, so that the component can be upgraded as before.
+
+On the other hand, the versioning mechanism should also be used by traits. For example, a trait can specify
+different revision and control deployment process. 
+
+
+## Goals
+
+1. Component is mutable. Developers who modify the Component object will directly trigger immediate
+upgrade of any `ApplicationConfiguration` that does not specify `Component` version. 
+2. Component Version can be specified in ApplicationConfiguration. Operator can specify a fixed version
+of Component in ApplicationConfiguration. Component mutation won't affect Application directly in this case, 
+and can be upgraded in a more convenient way by working with trait.
+3. The versioning mechanism should be compatible with current existing K8s workloads, eg, `K8s Deployment`, `OpenKruise`.
+  OAM should support two common ways used in community for versioning mechanism. 
+  - Workload itself can handle version. This means we can use `Deployment` only to do rollout or rollback without trait.
+  - Corporate with trait. OAM runtime should create different `Deployment` instances for different revision and let the trait do rollout or rollback.
+
+## Proposal
+
+### Using ControllerRevision to record versions of Component
+
+Because a `Component` is mutable, we introduce an independent Revision object to track a `Component`
+change automatically.
+
+We will use a new object which will be called as `ComponentRevision`, something like [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision) in K8s.
+
+For example, we have a following component yaml:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: frontend
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash lscpu"
+```
+
+When users apply this yaml file into the system, a `ComponentRevision` object will be created.
+
+```shell script
+$ kubectl apply -f component.yaml
+```
+
+```shell script
+$ kubectl get componentrevisions
+NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
+default       frontend-v0.0.1    core.oam.dev/component   1          2d15h
+
+$ kubectl get componentrevisions frontend-v0.0.1 -o yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ComponentRevision
+metadata:
+  name: frontend-v0.0.1 # you could name this anything you wanted, but just appending the semver would be good practice
+spec:
+  version: v0.0.1
```

Comment:
There's no need to explicitly claim a `version` IMO. Revision object should be generated by the system, so it doesn't make lots sense to force the system generate version number incrementally like this.

Reply:
I'm fine but what's the difference between controllerRevision with  ComponentRevision if we remove this version field ?

Reply:
The main difference is `ControllerRevision` is not human readable as it uses a generic `data` field to carry the revisioned object.

So if the argument is user experience, `ComponentRevision` proposed in https://github.com/crossplane/oam-kubernetes-runtime/pull/29#discussion_r431915520 is better - still, not a very big difference indeed. 

While what I am not sure on Daniel is comment is:

> but does not appear to be robust enough for a case like Component where an organization may want to "publish" multiple versions of a Component that will be consumed by multiple ApplicationConfigurations

I don't think it's the intention in this proposal. Revision is OAM is mostly for AppConfig/Trait controller to find workload CR by given revision number.